include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -0,0 +1,453 @@
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_MATH_FUNCTIONS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATH_FUNCTIONS_INCLUDED_
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include "nbl/builtin/hlsl/numbers.hlsl"
+#include "nbl/builtin/hlsl/spirv_intrinsics/core.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace math
+{
+
+namespace impl
+{
+template<typename T, uint32_t LP, bool Odd=(LP&0x1)>
+struct lp_norm;
+
+// infinity case
+template<typename T>
+struct lp_norm<T,0,false>
+{
+    static scalar_type_t<T> __call(const T v)
+    {
+        scalar_type_t<T> retval = abs<T>(v[0]);
+        for (int i = 1; i < rank<T>::value; i++)
+            retval = max<T>(abs<T>(v[i]),retval);
+        return retval;
+    }
+};
+
+// TOOD: is this doing what it should be?
+template<typename T>
+struct lp_norm<T,1,false>
+{
+    static scalar_type_t<T> __sum(const T v)
+    {
+        scalar_type_t<T> retval = abs<T>(v[0]);
+        for (int i = 1; i < rank<T>::value; i++)
+            retval += abs<T>(v[i]);
+        return retval;
+    }
+
+    static scalar_type_t<T> __call(const T v)
+    {
+        return __sum(v);
+    }
+};
+
+template<typename T>
+struct lp_norm<T,2,false>
+{
+    static scalar_type_t<T> __sum(const T v)
+    {
+        return dot<T>(v, v);   // TODO: wait for overloaded dot?
+    }
+
+    static scalar_type_t<T> __call(const T v)
+    {
+        return sqrt<T>(__sum(v));
+    }
+};
+
+// TODO: even/odd cases
+}
+
+template<typename T, uint32_t LP NBL_FUNC_REQUIRES(LP>0)
+scalar_type_t<T> lpNormPreroot(NBL_CONST_REF_ARG(T) v)
+{
+    return impl::lp_norm<T,LP>::__sum(v);
+}
+
+template<typename T, uint32_t LP>
+scalar_type_t<T> lpNorm(NBL_CONST_REF_ARG(T) v)
+{
+    return impl::lp_norm<T,LP>::__call(v);
+}
+
+
+template <typename T NBL_FUNC_REQUIRES(is_scalar_v<T>)
+vector<T, 3> reflect(vector<T, 3> I, vector<T, 3> N, T NdotI)
+{
+    return N * 2.0f * NdotI - I;
+}
+
+template <typename T NBL_FUNC_REQUIRES(is_scalar_v<T>)
+vector<T, 3> reflect(vector<T, 3> I, vector<T, 3> N)
+{
+    T NdotI = dot<T>(N, I);
```

Comment:
1. your `dot` wont compile, because its `dot<VectorT>`
2. I'd probably require `/*Vectorial<T> && */vector_traits<T>::Dimension==3` instead of taking `T` and slapping it in a `vector` (perf note, vector_traits existing is same as checking Vectorial but will make Clang/DXC choke a lot less)